enhancements/installer/connected-assisted-installer.md

```diff
@@ -0,0 +1,324 @@
+## Proposal
+
+There are several separate changes to enable the assisted installer
+workflows, including a GUI front-end for the installer, a cloud-based
+orchestration service, and changes to the installer and bootstrapping
+process.
+
+The process starts when the user goes to an "assisted installer"
+application running on `cloud.redhat.com`, enters details needed by
+the installer (OpenShift version, ssh keys, proxy settings, etc.), and
+then downloads a live RHCOS ISO image with the software and settings
+they need to complete the installation locally.
+
+The user then boots the live ISO on each host they want to be part of
+the cluster (control plane and workers). They can do this by hand
+using thumb drives, by attaching the ISO using virtual media support
+in the BMC of the host, or any other way they choose.
+
+When the ISO boots, it starts an agent that communicates with the REST
+API for the assisted installer service running on `cloud.redhat.com`
+to receive instructions. The agent registers the host with the
+service, using the user's pull secret embedded in the ISO's ignition config
+to authenticate. The agent identifies itself based on the serial
+number from the host it is running on. Communication always flows from
+agent to service via HTTPS so that firewalls and proxies work as
+expected.
+
+Each host agent periodically asks the service what tasks to perform,
+and the service replies with a list of commands and arguments. A
+command can be to:
+
+1. Return hardware information for its host
+2. Return L2 and L3 connectivity information between its host and the
+   other hosts (the IPs and MAC addresses of the other hosts are
+   passed as arguments)
+3. Begin the installation of its host (arguments include the host's
+   role, boot device, etc.). The agent executes different installation
+   logic depending on its role (bootstrap-master, master, or worker).
+
+The agent posts the results for the command back to the
+service. During the actual installation, the agents post progress.
+
+As agents report to the assisted installer, their hosts appear in the
+UI and the user is given an opportunity to examine the hardware
+details reported and to set the role and cluster of each host.
+
+The assisted installer orchestrates a set of validations on all
+hosts. It ensures there is full L2 and L3 connectivity between all of
+the hosts, that the hosts all meet minimum hardware requirements, and
+that the API and ingress VIPs are on the same machine network.
+
+The discovered hardware and networking details are combined with the
+results of the validation to derive defaults for the machine network
+CIDR, the API VIP, and other network configuration settings for the
+hosts.
+
+When enough hosts are configured, the assisted installer application
+replies to the agent on each host with the instructions it needs to
+take part in forming the cluster. The assisted installer application
+selects one host to run the bootstrap services used during
```

Comment:
In regard to the bootstrap service.
I think it's worth mentioning that unlike all other OpenShift installers, the assisted installer doesn’t use an auxiliary bootstrap node.
Instead, it will pivot the bootstrap to become a master node once the control plane is running on 2 other master nodes.
This flow reduces the minimum number of nodes to 3.

Reply:
> Instead, it will pivot the bootstrap to become a master node once the control plane is running on 2 other master nodes.

Do we run a descheduler by default now?  In addition to the "did we clean out all the bootstrap-specific stuff?" concerns we had about pivoting bootstrap into a compute node, pivoting into a control-plane node adds "do we end up with most of the critical pieces all lumped together on the two born-as-control-plane machines?".  Although I guess a few rounds of control-plane reboots during subsequent updates and config rollouts would wash away any initial lopsided balancing.

Reply:
>  In addition to the "did we clean out all the bootstrap-specific stuff?" concerns we had about pivoting bootstrap into a compute node, 

This is a pretty easy problem to solve, we can add support to Ignition to make everything ephemeral at the OS level (i.e. mount `/var` as a `tmpfs`, `/etc` as an overlayfs and everything else read-only).  Then rebooting guarantees everything done at the filesystem level is gone (and there's no reason for an OpenShift install process to operate at the block level).

Reply:
This part of the discussion should probably move to #361.